File: bank-scraper/domain/historic.py
from datetime import datetime, date
from typing import Optional

from pydantic.dataclasses import dataclass

from domain.transactions import ProductType, BaseInvestmentTx


@dataclass
class BaseHistoricEntry:
    name: str
    invested: float
    returned: Optional[float]
    currency: str
    currencySymbol: str
    lastInvestDate: datetime
    lastTxDate: datetime
    effectiveMaturity: Optional[datetime]
    netReturn: Optional[float]
    fees: Optional[float]
    retentions: Optional[float]
    interests: Optional[float]
    state: Optional[str]
    entity: str
    productType: ProductType
    relatedTxs: list[BaseInvestmentTx]


@dataclass
class FactoringEntry(BaseHistoricEntry):
    interestRate: float
    netInterestRate: float
    maturity: date
    type: str


@dataclass
class RealStateCFEntry(BaseHistoricEntry):
    interestRate: float
    months: int
    potentialExtension: Optional[int]
    type: str
    businessType: str


@dataclass
class Historic:
    entries: list[BaseHistoricEntry]

File: frontend/app/src/types/historic.ts
import type { Entity } from "."
import { DataSource } from "."
import { ProductType } from "./position"
import type { BaseInvestmentTx } from "./transactions"

export interface BaseHistoricEntry {
  id: string
  name: string
  invested: number
  repaid?: number | null
  returned?: number | null
  currency: string
  last_invest_date: string
  last_tx_date: string
  effective_maturity?: string | null
  net_return?: number | null
  fees?: number | null
  retentions?: number | null
  interests?: number | null
  state?: string | null
  entity: Entity
  product_type: ProductType
  related_txs: BaseInvestmentTx[]
  entity_account_id?: string | null
  source?: DataSource
  manual_key?: string | null
}

export interface FactoringEntry extends BaseHistoricEntry {
  interest_rate: number
  gross_interest_rate: number
  maturity: string
  type: string
}

export interface RealEstateCFEntry extends BaseHistoricEntry {
  interest_rate: number
  maturity: string
  extended_maturity?: string | null
  type: string
  business_type: string
}

export interface Historic {
  entries: BaseHistoricEntry[]
}

export type HistoricSortBy = "maturity" | "last_invest_date" | "invested"
export type SortOrder = "asc" | "desc"

export interface HistoricQueryRequest {
  entities?: string[]
  product_types?: ProductType[]
  page?: number
  limit?: number
  sort_by?: HistoricSortBy
  sort_order?: SortOrder
}

export interface SettleManualInvestmentRequest {
  entity_id: string
  entry_id: string
  product_type: ProductType
  maturity?: string | null
  interests?: number | null
  fees?: number
  retentions?: number
  pending_capital?: number
  create_investment_tx?: boolean
}

export interface PartialAmortizeManualInvestmentRequest {
  entity_id: string
  entry_id: string
  product_type: ProductType
  amount: number
  date?: string | null
  interests?: number
  fees?: number
  retentions?: number
  create_investment_tx?: boolean
}

export enum HistoricTxDeletion {
  NONE = "NONE",
  SETTLEMENT = "SETTLEMENT",
  ALL = "ALL",
}
